### DISCUSSION TITLE
How to create Netbox plugin?

### DISCUSSION THREAD

**Discussion by vladimirs-git** [netbox-community/netbox#14422] [Q&A] [answered]
I want to create a Netbox plugin. 
I have experience with Python, Netbox API, Flask, FastAPI, but have no experience with Django. How should I start? 
Please advise any useful resources, books, YouTube or courses to gain experience.

**Comment by mtinberg** [comment]
Honestly the Django project site documentation is excellent and has extensive tutorials

https://docs.djangoproject.com

in addition to any specific documentation for Netbox


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: Vladimir Prusakov ***@***.***>
Sent: Monday, December 4, 2023 10:12 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] How to create Netbox plugin? (Discussion #14422)


I want to create a Netbox plugin.
I have experience with Python, Netbox API, Flask, FastAPI, but have no experience with Django. How should I start?
Please advise any useful resources, books, YouTube or courses to gain experience.

**Comment by vladimirs-git** [comment]
I acknowledge that Django serves is the foundation, but I am mostly looking for tutorials that provide step-by-step guidance on creating Netbox plugin.

**Comment by arthanson** [comment, marked as answer]
There is pretty good documentation in NetBox docs: https://docs.netbox.dev/en/stable/plugins/development/

Here is a NetBox plugin development tutorial: https://github.com/netbox-community/netbox-plugin-tutorial

There is also a package cookiecutter-netbox-plugin (https://github.com/netbox-community/cookiecutter-netbox-plugin) that will scaffold a base plugin that you can extend that makes it easier to get started.

**Reply by candlerb** [reply]
In addition, the [netbox shell](https://docs.netbox.dev/en/stable/administration/netbox-shell/) (nbshell) is a good environment for interactively playing with the Django ORM.

Before you start on plugins, you might want to write a few [reports](https://docs.netbox.dev/en/stable/customization/reports/) or [custom scripts](https://docs.netbox.dev/en/stable/customization/custom-scripts/), both of which use the ORM but are simpler to get started with. There are examples of these in the [customizations](https://github.com/netbox-community/customizations) repository.
